### PR TITLE
Decouple loss computing and generation in GRPO

### PR DESCRIPTION
**Pull request by qgallouedec**
# What does this PR do?

Fixes # (issue)

## Before submitting
- [ ] This PR fixes a typo or improves the docs (you can dismiss the other checks if that's the case).
- [ ] Did you read the [contributor guideline](https://github.com/huggingface/trl/blob/main/CONTRIBUTING.md#create-a-pull-request),
      Pull Request section?
- [ ] Was this discussed/approved via a GitHub issue? Please add a link
      to it if that's the case.
- [ ] Did you make sure to update the documentation with your changes? Here are the
      [documentation guidelines](https://github.com/huggingface/trl/tree/main/docs).
- [ ] Did you write any new necessary tests?

## Who can review?

Anyone in the community is free to review the PR once the tests have passed. Feel free to tag
members/contributors who may be interested in your PR.

### CODE REVIEW

File: trl/trainer/grpo_trainer.py [hunk 1]
```diff
@@ -439,49 +451,28 @@ def compute_loss(self, model, inputs, return_outputs=False, num_items_in_batch=N
         completion_mask = (sequence_indices <= eos_idx.unsqueeze(1)).int()
 
         # Concatenate prompt_mask with completion_mask for logit computation
-        prompt_mask_repeated = prompt_inputs["attention_mask"].repeat_interleave(self.num_generations, dim=0)
-        attention_mask = torch.cat([prompt_mask_repeated, completion_mask], dim=1)  # (B*G, P+C)
-
-        # Get the per-token log probabilities for the completions for the model and the reference model
-        def get_per_token_logps(model, input_ids, attention_mask, logits_to_keep):
```
Comment (qgallouedec) [inline trl/trainer/grpo_trainer.py:446]: converted to method

File: trl/trainer/grpo_trainer.py [hunk 2]
```diff
@@ -439,49 +451,28 @@ def compute_loss(self, model, inputs, return_outputs=False, num_items_in_batch=N
...
-                with self.accelerator.unwrap_model(model).disable_adapter():
-                    ref_per_token_logps = get_per_token_logps(
-                        model, prompt_completion_ids, attention_mask, logits_to_keep
+                with self.accelerator.unwrap_model(self.model).disable_adapter():
+                    ref_per_token_logps = self._get_per_token_logps(
+                        self.model, prompt_completion_ids, attention_mask, logits_to_keep
                     )
 
-        # Compute the KL divergence between the model and the reference model
-        per_token_kl = torch.exp(ref_per_token_logps - per_token_logps) - (ref_per_token_logps - per_token_logps) - 1
```
Comment (qgallouedec) [inline trl/trainer/grpo_trainer.py:476]: this is later computed in `compute_loss`

File: trl/trainer/grpo_trainer.py [hunk 3]
```diff
@@ -521,15 +512,7 @@ def get_per_token_logps(model, input_ids, attention_mask, logits_to_keep):
...
         advantages = (rewards - mean_grouped_rewards) / (std_grouped_rewards + 1e-4)
 
-        # x - x.detach() allows for preserving gradients from x
-        per_token_loss = torch.exp(per_token_logps - per_token_logps.detach()) * advantages.unsqueeze(1)
-        per_token_loss = -(per_token_loss - self.beta * per_token_kl)
-        loss = ((per_token_loss * completion_mask).sum(dim=1) / completion_mask.sum(dim=1)).mean()
-
         # Log the metrics
-        completion_length = self.accelerator.gather_for_metrics(completion_mask.sum(1)).float().mean().item()
-        self._metrics["completion_length"].append(completion_length)
```
Comment (qgallouedec) [inline trl/trainer/grpo_trainer.py:531]: this is later computed in `compute_loss`

File: trl/trainer/grpo_trainer.py [hunk 4]
```diff
@@ -366,32 +366,41 @@ def _set_signature_columns_if_needed(self):
         if self._signature_columns is None:
             self._signature_columns = ["prompt"]
 
-    # Trainer "prepares" the inputs before calling `compute_loss`. It converts to tensor and move to device.
-    # Since we preprocess the data in `compute_loss`, we need to override this method to skip this step.
-    def _prepare_inputs(self, inputs: dict[str, Union[torch.Tensor, Any]]) -> dict[str, Union[torch.Tensor, Any]]:
```
Comment (qgallouedec) [inline trl/trainer/grpo_trainer.py:371]: in this method, we now:
- generate
- compute reward
- compute ref log probs

**Review by lewtun — APPROVED** [review]
Nice refactor, LGTM!

**Review by edbeeching — APPROVED** [review]
LGTM, looking forward to the next PR!
